File: dotnet/src/SimpleOrm/SimpleOrmException.cs
namespace SimpleOrm;

/// <summary>
/// A runtime failure with a stable error code from spec/errors.md (parameter
/// binding, query execution, transactions). Metadata loading uses
/// <see cref="MappingException"/>, which aggregates.
/// </summary>
public sealed class SimpleOrmException : Exception
{
    public SimpleOrmException(string code, string target, string message)
        : base($"{code} {target}: {message}")
    {
        Code = code;
        Target = target;
    }

    /// <summary>Stable code, e.g. <c>PRM-001</c> or <c>QRY-002</c>.</summary>
    public string Code { get; }

    /// <summary>What the error names: a query, parameter, or session.</summary>
    public string Target { get; }
}

File: php/src/Errors/SimpleOrmException.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Errors;

use RuntimeException;

/**
 * A failure with a stable error code from spec/errors.md (§2 "errors name
 * things"): the code is the cross-language contract, the target is the thing at
 * fault — a query, `Type.property`, a parameter, the session — and the message
 * says what was expected. Reads as "<code> <target>: <message>".
 *
 * Not final: ConcurrencyException and MappingException specialize it.
 */
class SimpleOrmException extends RuntimeException
{
    public function __construct(
        public readonly string $errorCode,
        public readonly string $target,
        string $message,
    ) {
        parent::__construct("{$errorCode} {$target}: {$message}");
    }
}
